java/address-java/src/main/java/uk/os/elements/address/utils/PostcodeUtil.java

```diff
@@ -0,0 +1,187 @@
+ * limitations under the License.
+ */
+
+package uk.os.elements.address.utils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.HashSet;
+import java.util.Scanner;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+/**
+ * Provide common utility functions for postcodes.
+ *
+ * <p>
+ * Note:
+ *
+ * <ul>
+ * <li>The Royal Mail is the data authority
+ * <ul>
+ * <li>they reserve the right to change their data at any time
+ * </ul>
+ * </li>
+ * <li>the only way to establish if a postcode is valid is to check the source postcode data
+ * <li>these utility functions are only checked against current postcodes; historical and future postcodes may fail these tests.
+ * </ul>
+ *
+ * <p>
+ * Breakdown of a postcode.
+ *
+ * <p>
+ * Royal Mail (2016, p.18) provide the following example:
+ *
+ * Using postcode "PO1 1AF" the components are:
+ * 'PO'1 1AF  - postcode area
+ *  PO'1' 1AF - postcode district
+ *  PO1 '1'AF - postcode sector
+ *  PO1 1'AF' - unit postcode
+ *
+ *  In the following checks we will assume component consists of all preceding components.
+ *
+ * <a href="http://www.royalmail.com/sites/default/files/docs/pdf/programmers_guide_edition_7_v5.pdf">Royal Mail (2016), Programmers' Guide</a>
+ */
+public final class PostcodeUtil {
+
+    private static final Set<String> POSTCODE_AREAS = getPostcodeAreas();
+
+    /**
+     * test if at least district level postcode
+     * // TODO agree definition with API, Data Delivery and Mobile teams
+     * @param value String under test
+     * @return true if the input is likely to be a postcode
+     */
+    public static boolean isLikelyPostcode(String value) {
+        return isLikelyDistrictPostcode(value) || isLikelySectorPostcode(value) || isLikelyFullPostcode(value);
+    }
+
+    /**
+     * @param value String under test
+     * @return true if the input is a full postcode candidate, false if partial or highly unlikely
+     */
+    public static boolean isLikelyFullPostcode(String value) {
+        return isLikelyUnitPostcode(value);
+    }
+
+    /**
+     * Assuming the postcode 'PO1 1AF', the area postcode is 'PO'
+     * @param value String under test
+     * @return true if considered an area postcode
+     */
+    public static boolean isLikelyAreaPostcode(String value) {
+        return POSTCODE_AREAS.contains(value.toUpperCase().trim());
+    }
+
+    /**
+     * Assuming the postcode 'PO1 1AF', the district postcode is 'PO1'
+     * @param value String under test
+     * @return true if considered a district postcode
+     */
+    public static boolean isLikelyDistrictPostcode(String value) {
+        Pattern pattern = Pattern.compile("^[A-IK-PR-UWYZ]?[A-H,K-Y][0-9]?[0-9A-HJKMNP-Y]$");
+        Matcher matcher = pattern.matcher(value.toUpperCase().trim());
+        return matcher.find() && value.matches(".*\\d.*");
+    }
+
+    public static boolean isLikelyDistrictPostcodeStrict(String value) {
```

Comment:
*my thought process*
1) ignore spaces and case: _most folk don't bother with a space in a postcode and they also probably don't care about case - I have often see modpostcode in databases with just the space removed for ease of matching.  I was thinking that util functions that are accepting of spaces would be helpful._

2) some might value space and possibly case: _I accept that spacing is useful to avoid ambiguity (esp. with districts / sectors) and could be useful to some devs.   I therefore created strict functions, which could also mandate case if that is deemed important.  re casing, I haven't seen anything to suggest that a valid postcode is uppercase - the PAF guide simply says all their data is uppercase._

Back to this point:
We don't need to be concerned with spacing here (it shouldn't have any) and I only used this method as an opportunity to pattern match against the postcode district format - which the (relaxed) regex doesn't do.  I also checked against data for a postcode area.

I see a couple of concerns from this work:
1) we can validate the format - which ignores all data and is what the Royal Mail publish
2) we can validate the data - by including data or a creating a regex that is updated as the data changes.

I have since updated the javadoc in a subsequent commit, which basically says that we check the format and the postcode area data.